services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelper.java

```diff
@@ -0,0 +1,208 @@
+import software.amazon.awssdk.utils.StringUtils;
+
+/**
+ * An internal helper class that traverses the file tree and send the upload request
+ * for each file.
+ */
+@SdkInternalApi
+public class UploadDirectoryHelper {
+    private static final Logger log = Logger.loggerFor(S3TransferManager.class);
+    private static final String DEFAULT_DELIMITER = "/";
+
+    private final TransferManagerConfiguration transferConfiguration;
+    private final Function<UploadRequest, Upload> uploadFunction;
+    private final FileSystem fileSystem;
+
+    public UploadDirectoryHelper(TransferManagerConfiguration transferConfiguration,
+                                 Function<UploadRequest, Upload> uploadFunction) {
+
+        this.transferConfiguration = transferConfiguration;
+        this.uploadFunction = uploadFunction;
+        this.fileSystem = FileSystems.getDefault();
+    }
+
+    @SdkTestInternalApi
+    UploadDirectoryHelper(TransferManagerConfiguration transferConfiguration,
+                          Function<UploadRequest, Upload> uploadFunction,
+                          FileSystem fileSystem) {
+
+        this.transferConfiguration = transferConfiguration;
+        this.uploadFunction = uploadFunction;
+        this.fileSystem = fileSystem;
+    }
+
+    public UploadDirectoryTransfer uploadDirectory(UploadDirectoryRequest uploadDirectoryRequest) {
+
+        CompletableFuture<CompletedUploadDirectory> returnFuture = new CompletableFuture<>();
+
+        // offload the execution to the transfer manager executor
+        CompletableFuture.runAsync(() -> doUploadDirectory(returnFuture, uploadDirectoryRequest),
+                                   transferConfiguration.option(TransferConfigurationOption.EXECUTOR));
+
+        return UploadDirectoryTransfer.builder().completionFuture(returnFuture).build();
+    }
+
+    private void doUploadDirectory(CompletableFuture<CompletedUploadDirectory> returnFuture,
+                                   UploadDirectoryRequest uploadDirectoryRequest) {
+
+        Path directory = uploadDirectoryRequest.sourceDirectory();
+        List<FailedSingleFileUpload> failedUploads = Collections.synchronizedList(new ArrayList<>());
+        Phaser phaser = new Phaser(1);
+
+        try (Stream<Path> entries = listFiles(directory, uploadDirectoryRequest)) {
+            entries.forEach(path -> {
+                CompletableFuture<CompletedUpload> future =
+                    uploadSingleFile(uploadDirectoryRequest, failedUploads, path, phaser);
+
+                // Forward cancellation of the return future to all individual futures.
+                CompletableFutureUtils.forwardExceptionTo(returnFuture, future);
+            });
+        }
+
+        // Wait for all sub transfers to complete
+        phaser.arriveAndAwaitAdvance();
+        phaser.arriveAndDeregister();
+        returnFuture.complete(CompletedUploadDirectory.builder().failedUploads(failedUploads).build());
```

Comment:
Instead of using a phaser and blocking this thread, should this thread just be used for delegation, and then we can return `CompletableFuture.allOf(uploadSingleFileFutures)`?

That'll make this thread's work much shorter-lived, and make this executor less likely to be a bottleneck.